Split deepvariant 3

fa2k opened the pull request:
<!--
# nf-core/modules pull request

Many thanks for contributing to nf-core/modules!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the master branch.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

Closes https://github.com/nf-core/modules/issues/5793

This is the third attempt at splitting DeepVariant. The goal is to make it possible to run call_variants on GPU whilst not allocating GPU resources for the CPU-only pipeline stages make_examples and postprocess_variants.

I needed to start over to base it on the new version of DeepVariant from the master branch, and to keep the existing module: The previous PR removed the module and replaced it with a subworkflow. This PR renames the existing module from `deepvariant` to `deepvariant/rundeepvariant`, and adds the subworkflow and subcommands used by it. It allows pipeline to choose between using an integrated single module, and a subworkflow with three processes.

Making a new PR unfortunately messes with @fellen31's review (in the previous attempt #5990), so I'll try to recreate the review items that weren't addressed yet.


- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [x] `nf-core modules test <MODULE> --profile singularity`
    - ❌ `nf-core modules test <MODULE> --profile conda` NOT supported
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [x] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - ❌ `nf-core subworkflows test <SUBWORKFLOW> --profile conda` NOT supported

--- 

Fixed general issues:

* Removed unique hash from makeexamples, instead relying on meta uniqueness.
* Upgraded to DeepVariant 1.6.1

fa2k commented on modules/nf-core/deepvariant/makeexamples/main.nf:
```diff
@@ -0,0 +1,64 @@
+        --ref "${fasta}" \\
+        --reads "${input}" \\
+        --examples "./${prefix}.examples.tfrecord@${task.cpus}.gz" \\
+        --channels "insert_size" \\
```
https://github.com/nf-core/modules/pull/5990#discussion_r1687551686

TODO

fellen31 replied:
> [fellen31](https://github.com/fellen31) [3 weeks ago](https://github.com/nf-core/modules/pull/6172#discussion_r1687551686)
Suggested change
        --channels "insert_size" \\
This needs to go into config, since it's not active for HYBRID_PACBIO_ILLUMINA, PACBIO or ONT_R104 model types.
Member
Author
@[fa2k](https://github.com/fa2k) fa2k [yesterday](https://github.com/nf-core/modules/pull/6172#discussion_r1715335472)
Yes according to the guidelines, the arguments to the tool should be minimal and anything extra put in the config. So I agree with this, but I fear that leaving out insert_size could make it very easy to misconfigure the tool for users. What could be done to help users remember to apply this channel using config when appropriate?

> Do you reckon this should go into ext.args or should I create a dedicated config variable for insert size?

I see your point, the wgs/wes models will be used most, but by hardcoding `--channels "insert_size"` in the module  you will not be able to run the long-read/hybrid models at all. 

Who are the users here? As I see it, a pipeline will have this subworkflow with a config. The subworkflow takes a val(model_type), and the settings for make_examples is determined by this, either by logic in the subworkflow config, or model_type is passed on to makeexamples and the logic is in the module. The burden is either pushed a bit more towards the pipeline maintainers, or the module maintainers. 

> What could be done to help users remember to apply this channel using config when appropriate?

I can see the point of having the logic in the module, even though it might stray from the guidelines. 

The end user of a pipeline should still be able to use a simple parameter or the samplesheet  to choose between wes or wgs analysis.  

> Do you reckon this should go into ext.args or should I create a dedicated config variable for insert size?

If you go with having the logic in config, I think the consensus is to use `ext.args`, and not have a dedicated config variables.

fellen31 replied:
By keeping the current deepvariant as rundeepvariant as well, pipeline maintainers still have the option to go with the simple solution where you don't need to care about knowing the correct makeexamples parameters for a certain model.

fa2k replied:
I've removed --channels from makeexamples, and adding a note to the README in modules/nf-core/deepvariant. For other data types it may be more complex to see what channels are needed.

I kind of agree with your reasoning that it's the pipeline maintainer's responsibility. (As a pipeline user I personally find the pipeline's config dir difficult to read, when trying to understand a pipeline, but I'm slowly learning to always look in the config.)

fa2k commented on modules/nf-core/deepvariant/callvariants/main.nf:
```diff
@@ -0,0 +1,61 @@
+
+    input:
+    tuple val(meta), path(make_examples_tfrecords)
+    val model_type
```
https://github.com/nf-core/modules/pull/5990#discussion_r1687554125
fellen31:
> We will need to deal with model_type for makeexamples in the config. Maybe we should remove model_type here, and keep that logic in the config as well.

fa2k replied:
Model type (--checkpoint arg) moved to config

fa2k commented on modules/nf-core/deepvariant/makeexamples/main.nf:
```diff
@@ -0,0 +1,64 @@
+    tuple val(meta), path(input), path(index), path(intervals)
+    tuple val(meta2), path(fasta)
+    tuple val(meta3), path(fai)
+    tuple val(meta4), path(gzi)
```
> Is this tuple val(meta4), path(gzi) needed?

TBC

fellen31 replied:
Think you can disregard my comment on this.

fa2k replied:
Ok I will leave the option for gzi to make it easy :)  I didn't get a chance to test whether it works.

fa2k commented on subworkflows/nf-core/deepvariant/main.nf:
```diff
@@ -0,0 +1,45 @@
+    ch_fasta            // channel: [ val(meta2), path(fasta) ]
+    ch_fai              // channel: [ val(meta3), path(fail) ]
+    ch_gzi              // channel: [ val(meta4), path(gzi) ]
+    ch_model_type       // channel: val("wgs" | "wes")
```
`ch_model_type       // channel: val("wgs" | "wes" | "ont_r104" | "pacbio" | "hybrid_pacbio_illumina" )`

TODO

fa2k replied:
I end up removing this, as the model will be specified using the config

fa2k commented on subworkflows/nf-core/deepvariant/tests/main.nf.test:
```diff
@@ -0,0 +1,104 @@
+    tag "deepvariant/callvariants"
+    tag "deepvariant/postprocessvariants"
+
+    test("homo_sapiens - two inputs - bam - fasta - fai") {
```
> This test and the one below only runs test and not test2. Not sure if removing Channel.of, or adding collect() to the rest of the inputs is the way to go.

fa2k commented on modules/nf-core/deepvariant/postprocessvariants/main.nf:
```diff
@@ -0,0 +1,77 @@
+process DEEPVARIANT_POSTPROCESSVARIANTS {
+    tag "$meta.id"
+    label 'process_medium'
```
The suggestion was to change from `process_single` to  `process_high` because the new Deepvariant postprocessvariants is multithreaded. I've changed to `process_medium` because it seems only part of it is multithreaded

fellen31 replied:
Sounds reasonable!

fa2k commented on modules/nf-core/deepvariant/callvariants/main.nf:
```diff
@@ -0,0 +1,61 @@
+    val model_type
+
+    output:
+    tuple val(meta), path("${prefix}.call-*-of-*.tfrecord.gz"), emit: call_variants_tfrecords
```
Removed val("${task.hash}") as it's not needed when meta is unique

fa2k replied:
I will add this information to the subworkflow's readme

fa2k reviewed (commented):
I'm just copying in @fellen31 's comments from  #5990 so they can be addressed in the new branch

fa2k commented:
This is now a working pipeline, but it is not necessarily ready for merging - see the issue https://github.com/nf-core/modules/issues/5793 for discussion

fellen31 commented:
Could you resolve the branch conflicts @fa2k and I'll review?

fa2k commented:
I'm working on this now, to add this new option for PAR bed file. I have a question, what would make sense to set for author and maintainer for the new modules and subworkflows? I kept the authors of the original deepvariant module, but I also added myself as author. I don't know if I should be "maintainer" or not?

fa2k commented:
Now the dev branch is merged and the new functionality is also integrated in the split deepvariant version.

fellen31 commented on subworkflows/nf-core/deepvariant/README.md:
```diff
@@ -0,0 +1,11 @@
+It is very important that the input channel's `meta` is unique for all the input elements, because the pipeline does
+a join on `meta`.
+
+Please note the important configuration items listed in the `deepvariant` module's README file. It is required to use the
+configuration to specify the input "channels" (data types to extract from bam file) for `DEEPVARIANT_MAKEEXAMPLES`, and the
+model to run for `DEEPVARIANT_CALLVARIANTS`.
```
```suggestion
Please note the important configuration items listed in the `deepvariant` module's README file. It is required to use the configuration to specify the input "channels" (data types to extract from bam file) for `DEEPVARIANT_MAKEEXAMPLES`, and the model to run for `DEEPVARIANT_CALLVARIANTS`.
```

fa2k commented on modules/nf-core/deepvariant/makeexamples/tests/main.nf.test:
```diff
@@ -0,0 +1,189 @@
+            assertAll(
+                { assert process.success },
+                // Test string values and versions. The make_examples_id and tfrecords are not stable.
+                { assert process.out.examples.get(0).get(0) == [ id:'test', single_end:false ] },
```
Yes this would be an improvement. The current test already assumes that there are two cpus in the process. I hope this is okay. @fellen31 do you know if there's a way to snapshot it with nf-test? Otherwise I can of course give a list of the correct names in the test code.

fellen31 replied:
The tests would fail then if someone else runs with another number of cpus, let's keep it the way you wrote it.

fa2k replied:
The tests I have include things like

                { assert process.out.examples.get(0).get(1).size() == 4 },

So it's checking for 4 files, which is based on 2 cpus. So it does need to change. I have to see if I can reference task.cpus in the "then" code, then I can compute the correct list of file names.

fellen31 replied:
Hm, maybe you could just check if `process.out.examples` has some files, but not snapshot them... not sure.

fa2k replied:
Can I force the number of cpus in the config?

fellen31 replied:
The tests should always work with the GitHub runners I guess, it's just if someone would test module locally. But I don't think that's very likely. It's not perfect, but maybe we can just keep it like it is.

fa2k replied:
Right, I agree, but I can change it to check for a list of filenames. There is no downside, since both tests make an assumption of the number of cpus.

fellen31 replied:
Sounds good! Think something like this might work, since you have two files: 

```
{ assert snapshot(
    file(process.out.bam.get(0).get(1)).name,
    file(process.out.bam.get(1).get(1)).name,
).match() }
```

fa2k replied:
Fixed now -cheking the file names

fellen31 commented:
Looks like one of the subworkflows tests are failing, maybe just need to update the snapshot?

fellen31 reviewed (approved):
I think this looks okay to me, maybe we can get a final OK from someone who uses deepvariant within nf-core, @maxulysse or @ramprasadn?

maxulysse commented:
Can you update https://github.com/nf-core/modules/blob/master/.github/workflows/test.yml#L541-L542 to ensure CI is not run for conda (as this modules is not working on it anyway)

fa2k commented:
Hm, now the singularity tests are failing

fa2k commented:
The failures could maybe be some infrastructure issue, as it doesn't seem to be related to my changes. Maybe I can try to run the tests again later.

maxulysse commented:
Yeah, I was about to suggest that

maxulysse commented:
I relaunched the failed ones

maxulysse commented:
Can you follow [this guide](https://nf-co.re/docs/guidelines/components/module_deprecation) and add a deprecation for the old module?

fa2k commented:
It's a very good idea to add a deprecation warning, and I've added back `main.nf` in the `deepvariant` module directory, and added a test that checks that this module fails.

fellen31 commented:
Good to merge? 😀
